src/hazelcore/extractions/ocaml/OCamlExtraction_Exp.re

```diff
@@ -0,0 +1,270 @@
+        ++ " : "
+        ++ ocaml_ht
+        ++ ") -> ("
+        ++ fst(ocaml_de)
+        ++ "))",
+        Arrow(ht, snd(ocaml_de)),
+      );
+    | _ => failwith("Exp: Lambda not variable")
+    }
+  | Ap(de1, de2) =>
+    //apply
+    let ocaml_de1 = extract(~ctx, ~de=de1);
+    let ocaml_de2 = extract(~ctx, ~de=de2);
+    switch (snd(ocaml_de1)) {
+    | Arrow(t1, t2) =>
+      // Hole -> some type should be ok
+      if (t1 == snd(ocaml_de2) || t1 == Hole) {
+        ("(" ++ fst(ocaml_de1) ++ " " ++ fst(ocaml_de2) ++ ")", t2);
+      } else {
+        failwith("Exp: Apply, type inconsistent");
+      }
+    | _ => failwith("Exp: Apply, not an Arrow type")
+    };
+  | BoolLit(bool) => (string_of_bool(bool), Bool)
+  | IntLit(int) => (string_of_int(int), Int)
+  | FloatLit(float) => (string_of_float(float), Float)
+  | BinBoolOp(op, d1, d2) =>
+    switch (d1, d2) {
+    | (BoolLit(b1), BoolLit(b2)) =>
+      let str_of_op: DHExp.BinBoolOp.t => string = (
+        fun
+        | And => " && "
+        | Or => " || "
+      );
+      // add parenthesis to avoid evaluation priority
+      let str =
+        "("
+        ++ string_of_bool(b1)
+        ++ str_of_op(op)
+        ++ string_of_bool(b2)
+        ++ ")";
+      (str, Bool);
+    | _ => failwith("Exp: BinBoolOp")
+    }
+  | BinIntOp(op, d1, d2) =>
+    let str_of_op: DHExp.BinIntOp.t => (string, HTyp.t) = (
+      fun
+      | Minus => (" - ", Int)
+      | Plus => (" + ", Int)
+      | Times => (" * ", Int)
+      | Divide => (" / ", Int)
+      | LessThan => (" < ", Bool)
+      | GreaterThan => (" > ", Bool)
+      | Equals => (" == ", Bool)
+    );
+    let ocaml_op = str_of_op(op);
+    let ocaml_d1 = extract(~ctx, ~de=d1);
+    let ocaml_d2 = extract(~ctx, ~de=d2);
+    switch (snd(ocaml_d1), snd(ocaml_d2)) {
+    | (Int, Int) => (
+        "(" ++ fst(ocaml_d1) ++ fst(ocaml_op) ++ fst(ocaml_d2) ++ ")",
+        snd(ocaml_op),
+      )
+    | _ => failwith("Exp: BinIntOp")
+    };
+  | BinFloatOp(op, d1, d2) =>
+    let str_of_op: DHExp.BinFloatOp.t => (string, HTyp.t) = (
+      fun
+      | FPlus => (" +. ", Float)
+      | FMinus => (" -. ", Float)
+      | FTimes => (" *. ", Float)
+      | FDivide => (" /. ", Float)
+      | FLessThan => (" < ", Bool)
+      | FGreaterThan => (" > ", Bool)
+      | FEquals => (" == ", Bool)
+    );
+    let ocaml_op = str_of_op(op);
+    let ocaml_d1 = extract(~ctx, ~de=d1);
+    let ocaml_d2 = extract(~ctx, ~de=d2);
+    switch (snd(ocaml_d1), snd(ocaml_d2)) {
+    | (Float, Float) => (
+        "(" ++ fst(ocaml_d1) ++ fst(ocaml_op) ++ fst(ocaml_d2) ++ ")",
+        snd(ocaml_op),
+      )
+    | _ => failwith("Exp: BinFloatOp")
+    };
+  | ListNil(t) => ("[]", List(t))
+  | Cons(d1, d2) =>
+    //TODO: add proper parenthesis to avoid priority in case "map"
+    let ocaml_d1 = extract(~ctx, ~de=d1);
+    let ocaml_d2 = extract(~ctx, ~de=d2);
+    switch (snd(ocaml_d2)) {
+    //FIXME: will ever happen? test!
+    | Hole => (
```

Comment:
won't ever happen (there would always be a cast), so don't need to handle cases like this